internal/xds/translator/extauth.go

```diff
@@ -0,0 +1,339 @@
+)
+
+func init() {
+	registerHTTPFilter(&extAuth{})
+}
+
+type extAuth struct {
+}
+
+var _ httpFilter = &extAuth{}
+
+// patchHCM builds and appends the ext_authz Filters to the HTTP Connection Manager
+// if applicable, and it does not already exist.
+// Note: this method creates an ext_authz filter for each route that contains an ExtAuthz config.
+func (*extAuth) patchHCM(mgr *hcmv3.HttpConnectionManager, irListener *ir.HTTPListener) error {
+	var errs error
+
+	if mgr == nil {
+		return errors.New("hcm is nil")
+	}
+
+	if irListener == nil {
+		return errors.New("ir listener is nil")
+	}
+
+	for _, route := range irListener.Routes {
+		if !routeContainsExtAuth(route) {
+			continue
+		}
+
+		filter, err := buildHCMExtAuthFilter(route)
+		if err != nil {
+			errs = errors.Join(errs, err)
+			continue
+		}
+
+		// skip if the filter already exists
+		for _, existingFilter := range mgr.HttpFilters {
+			if filter.Name == existingFilter.Name {
+				continue
+			}
+		}
+
+		mgr.HttpFilters = append(mgr.HttpFilters, filter)
+	}
+
+	return nil
+}
+
+// buildHCMExtAuthFilter returns a ext_authz HTTP filter from the provided IR HTTPRoute.
+func buildHCMExtAuthFilter(route *ir.HTTPRoute) (*hcmv3.HttpFilter, error) {
+	extAuthProto := extAuthConfig(route.ExtAuth)
+	if err := extAuthProto.ValidateAll(); err != nil {
+		return nil, err
+	}
+
+	extAuthAny, err := anypb.New(extAuthProto)
+	if err != nil {
+		return nil, err
+	}
+
+	return &hcmv3.HttpFilter{
+		Name: extAuthFilterName(route),
+		ConfigType: &hcmv3.HttpFilter_TypedConfig{
+			TypedConfig: extAuthAny,
+		},
+	}, nil
+}
+
+func extAuthFilterName(route *ir.HTTPRoute) string {
+	return perRouteFilterName(extAuthFilter, route.Name)
+}
+
+func extAuthConfig(extAuth *ir.ExtAuth) *extauthv3.ExtAuthz {
+	config := &extauthv3.ExtAuthz{}
+
+	var headersToExtAuth []*matcherv3.StringMatcher
+	for _, header := range extAuth.HeadersToExtAuth {
+		headersToExtAuth = append(headersToExtAuth, &matcherv3.StringMatcher{
+			MatchPattern: &matcherv3.StringMatcher_Exact{
+				Exact: header,
+			},
+		})
+	}
+
+	if len(headersToExtAuth) > 0 {
+		config.AllowedHeaders = &matcherv3.ListStringMatcher{
+			Patterns: headersToExtAuth,
+		}
+	}
+
+	if extAuth.HTTP != nil {
+		config.Services = &extauthv3.ExtAuthz_HttpService{
+			HttpService: httpService(extAuth.HTTP),
+		}
+	} else if extAuth.GRPC != nil {
+		config.Services = &extauthv3.ExtAuthz_GrpcService{
+			GrpcService: &corev3.GrpcService{
+				TargetSpecifier: &corev3.GrpcService_EnvoyGrpc_{
+					EnvoyGrpc: grpcService(extAuth.GRPC),
+				},
+				Timeout: &duration.Duration{
+					Seconds: defaultExtServiceRequestTimeout,
+				},
+			},
+		}
+	}
+
+	return config
+}
+
+func httpService(http *ir.HTTPExtAuthService) *extauthv3.HttpService {
+	var (
+		enableTLS        = http.TLS != nil
+		uri              string
+		headersToBackend []*matcherv3.StringMatcher
+	)
+
+	if enableTLS {
+		uri = fmt.Sprintf("https://%s:%d", http.Host, http.Port)
```

Comment:
Instead of building `uri` up as a string, why not use `net/url`'s [URL](https://pkg.go.dev/net/url#URL).

Code would look something like this:

```golang
uri := net.URL{
    Host:  http.Host,
    Port:  http.Port,
    Path:  ptr.Deref(http.Path, ""),
}
if enableTLS {
    url.Scheme = "https"
} else {
    url.Scheme = "http"
}
// Convert to a string for usage latter on.
urlAsString := uri.String()
```

This has the following advantages:
1. Properly escaping elements in the path
2. Making sure that a valid URL is created in corner cases. Example - the current code fails to create a valid URL if the `*http.Path` doesn't start with a `/`.